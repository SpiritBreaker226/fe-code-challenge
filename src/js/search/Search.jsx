import React from 'react';
import PropTypes from 'prop-types';

import {connect} from 'react-redux';

import {push} from 'connected-react-router';

import {updateSelected} from '../spot/spot-actions';
import SpotList from './spot-list/SpotList';
import SpotDetails from '../spot/SpotDetails';

const Search = ({
    selectedSpot,
    spots,
    setSpot,
    pushTo
}) => {
    return (
        <div className="Search">
            <SpotList
                spots={spots}
                selectedSpot={selectedSpot}
                setSpot={setSpot}
            />
            <div className="Search-content">
                {
                    selectedSpot &&
                        <SpotDetails
                            selectedSpot={selectedSpot}
                            setSpot={setSpot}
                            pushTo={pushTo}
                        />
                }
            </div>
        </div>
    );
};

Search.propTypes = {
    selectedSpot: PropTypes.object,
    spots: PropTypes.arrayOf(PropTypes.object).isRequired,
    setSpot: PropTypes.func.isRequired,
    pushTo: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
    const {
        spot: {
            selected: selectedSpot
        }
    } = state;

    return {
        selectedSpot
    };
};

const mapDispatchToProps = {
    setSpot: updateSelected,
    pushTo: push,
};

export default connect(mapStateToProps, mapDispatchToProps)(Search);
