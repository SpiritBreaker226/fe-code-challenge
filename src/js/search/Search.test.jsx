import React from 'react';
import {fireEvent, render, screen, waitFor} from '@testing-library/react';

import {Provider} from 'react-redux';

import axios from 'axios';

import spots from '../helpers/test_data_spots.json';

import createStore from '../store/store';

import Search from './Search';

jest.mock('axios');
const mockedAxios = axios;

describe('Search', () => {
    const elementToRender = testId => {
        return () => screen.queryByTestId(testId);
    };

    beforeEach(async () => {
        const mockStore = createStore(
            {
                spot: {
                    selected: null
                },
            },
        );

        render(
            <Provider store={mockStore}>
                <Search spots={spots} />
            </Provider>
        );

        await waitFor(() => screen.getByTestId('spot-item-1'));
    });

    it('renders Search component', () => {
        expect(screen.getByTestId('search')).toMatchSnapshot();
    });

    describe('on spot details', () => {
        it('render different Spot Details when clicking Details', async () => {
            mockedAxios.get.mockResolvedValue({data: spots[0]});

            fireEvent.click(screen.getByTestId('spot-item-details-button-1'));

            expect(mockedAxios.get).toHaveBeenCalledTimes(1);
            expect(mockedAxios.get).toHaveBeenCalledWith('/spots/1');

            await waitFor(elementToRender('spot-details'));

            expect(screen.getByTestId('spot-details-title')).toHaveTextContent(
                /Seven Lions/i
            );
            expect(screen.getByTestId('spot-details-purchase')).toHaveTextContent(
                /14.00/i
            );

            mockedAxios.get.mockResolvedValue({data: spots[1]});

            fireEvent.click(screen.getByTestId('spot-item-details-button-2'));

            expect(mockedAxios.get).toHaveBeenCalledWith('/spots/2');

            await waitFor(elementToRender('spot-details'));

            expect(screen.getByTestId('spot-details-title')).toHaveTextContent(
                /Garage/i
            );
            expect(screen.getByTestId('spot-details-purchase')).toHaveTextContent(
                /16.50/i
            );

            fireEvent.click(screen.getByTestId('spot-details-close'));

            expect(screen.queryByTestId('spot-details')).not.toBeInTheDocument();
        });
    });
});
