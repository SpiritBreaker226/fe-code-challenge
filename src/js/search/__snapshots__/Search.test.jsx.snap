// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders Search component 1`] = `
<div
  class="Search"
  data-testid="search"
>
  <div
    class="SpotList"
  >
    <div
      class="SpotList-feature"
    >
      <div
        class="SpotList-breadcrumbs"
      >
        <button
          class="TextButton"
          type="button"
        >
          Chicago
        </button>
         &gt; Millennium Park
      </div>
      <h1>
        Millennium Park
      </h1>
      <p>
        3
         Spots Available
      </p>
    </div>
    <div
      class="SpotList-spots"
    >
      <div
        class="SpotItem"
        data-testid="spot-item-1"
      >
        <img
          alt=""
          class="Image"
          src="https://res.cloudinary.com/spothero/image/upload/v1579704583/front-end/code-challenge/wkabs86wl5zwqaksb6fr.jpg"
        />
        <div
          class="SpotItem-info"
          data-testid="spot-item-info-1"
        >
          <h2>
            Seven Lions - Valet
          </h2>
          <p>
            0.1 mi
          </p>
          <button
            class="TextButton"
            data-testid="spot-item-details-button-1"
            type="button"
          >
            Details
          </button>
        </div>
      </div>
      <div
        class="SpotItem"
        data-testid="spot-item-2"
      >
        <img
          alt=""
          class="Image"
          src="https://res.cloudinary.com/spothero/image/upload/v1579704583/front-end/code-challenge/vlv2pdojzj3yf8bdtuft.jpg"
        />
        <div
          class="SpotItem-info"
          data-testid="spot-item-info-2"
        >
          <h2>
            55 E Monroe St. - Garage
          </h2>
          <p>
            390 ft
          </p>
          <button
            class="TextButton"
            data-testid="spot-item-details-button-2"
            type="button"
          >
            Details
          </button>
        </div>
      </div>
      <div
        class="SpotItem"
        data-testid="spot-item-3"
      >
        <img
          alt=""
          class="Image"
          src="https://res.cloudinary.com/spothero/image/upload/v1579704583/front-end/code-challenge/gkcgzedhqojsrobzdvsa.jpg"
        />
        <div
          class="SpotItem-info"
          data-testid="spot-item-info-3"
        >
          <h2>
            124 S Wabash Ave. - Garage
          </h2>
          <p>
            0.1 mi
          </p>
          <button
            class="TextButton"
            data-testid="spot-item-details-button-3"
            type="button"
          >
            Details
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="Search-content"
  />
</div>
`;
